### DISCUSSION TITLE
Should libsidplayfp be made thread-safe?

### DISCUSSION THREAD

**Discussion by Toddler-Boy** [libsidplayfp/libsidplayfp#89] [General]
If we could identify the problematic functions/structures/code-segments that prevent this, is there even a desire to fix it?

I would like to see it happening, as I could think of a few use cases where it would make the users' lives a lot easier. Maybe we can use ASAN to identify the problematic areas and then tackle them one by one?

**Comment by drfiemost** [comment]
As mentioned in #71 there are a few singletons and static structures used as cache that need to be protected with mutexes or whatever.
I'm not really motivated to work on it but contributions are surely welcome.

**Comment by Toddler-Boy** [comment]
Yes, but in order for others, like myself, to help, we need to know all the places this might happen. If you know which files and functions are affected, please let us know. std::mutex has been around since c++11, so it could be done safely and easily without affecting compatibility.

**Comment by Toddler-Boy** [comment]
So, I quickly searched for "static", and fixing this issue looks very straightforward. Before I begin, I would like to clarify a few things:

1) Why are these things static in the first place? They are small(ish) tables that need only 64kb or 128kb. If a system is so memory-restricted that it can't afford to spend 128kb on a couple of tables, then it probably also isn't performant enough to use this library at all. It seems this was done "just cause".

2) I didn't analyze the whole codebase, but wouldn't the adjustable filter curve also update these tables? If so, then having them static is impossible unless all threads use the exact same settings.

3) The changes to protect/remove static variables wouldn't work for the hardsid and exsid implementation, as people usually only have one hardware, and their statics are used differently there anyway.

4) The OPEN_MP pragmas throw several warnings, because they are used even if OPEN_MP is not used. Since the tables these functions calculate are relatively small, is the OPEN MP stuff even needed at all? Was the initialization performance even noticeable before the change? I remember starting with my player before this change and I have OPEN MP still turned off, and the initialization feels instant to me.

5) Since this fork has diverged a bit from the original reSID implementation by Dag Lem, are there any plans to include the more recent changes into this fork as well? It seems there has been a lot of activity since 2021 to improve accuracy and get rid of sampled waveforms in favor of real-time calculated ones.

I hope I'm not rocking the boat too much. I just want to help make this library better.

**Reply by drfiemost** [reply]
1. The big problems are the filter tables:
https://github.com/libsidplayfp/libsidplayfp/blob/b0ceb7672325c1025b5ae7aa28cd4d5afc3d3f77/src/builders/residfp-builder/residfp/FilterModelConfig6581.h#L48
and the resampling tables, which change depending on the sampling rate and are cached in a static var:
https://github.com/libsidplayfp/libsidplayfp/blob/b0ceb7672325c1025b5ae7aa28cd4d5afc3d3f77/src/builders/residfp-builder/residfp/resample/SincResampler.cpp#L51
These should be initializated by the first thread while other threads wait for the initialization to complete.
Other static structures like the envelope table are fixed at compile time and pose no problem.

2. IIRC the filter tables parameters are fixed at compile time.

3. Yes, I guess it wouldn't make sense to multithread hardware devices playback.

4. The filter tables are a few MBs. There is a small but noticeable speedup a startup for me, maybe because my processor is quite dated. What kind of warnings do you see?

5. Not sure what you mean, reSIDfp is a fork with "experimental" features like compile-time calculated combined waveforms and two-step resampling. All changes that have been proved good was merged back into reSID, currently maintained as part of VICE. So both emulations are quality-wise excellent with just some minor differences.

**Comment by Toddler-Boy** [comment]
1. So for the filter and resample tables, a simple mutex would be fine, but for the re-sampling table I would actually switch to a locally owned version, as this allows the sample-rate to be different between threads. This can be useful for exporting a e.g. 48 kHz WAV-file, while listening to a different tune at 44.1 kHz. In today's world, with multiple gigabyes of memory, a couple of MB is literally nothing. My own machine has 64 GB. Gamers have at least 16 GB, now shifting to 32 GB. Expecting less than 4 GB on a machine in use today is unrealistic. Even if your machine had only 1 GB, spending 2 MB on a table is only 0.2% of that available memory. Nobody would care.

We also have to remember: as long as you only have one thread (and more is not even possible right now), the memory consumption would be the exact same. So what is there to lose? Making all these variables non-static doesn't lose anything. You're not spending any more memory, start-up time, or anything else. It stays the same. I truly don't understand why they are static in the first place. It provides zero benefits, at the cost of multi-threaded compatibility.

> The filter tables are a few MBs. There is a small but noticeable speedup a startup for me, maybe because my processor is quite dated. What kind of warnings do you see?

```-Wsource-uses-openmp``` gets thrown by clang. It's because you've used ```#pragma omp``` outside the guards.

I can measure performance, but even in debug mode (which on my machine is more than 10x as slow as release, I don't notice any initialization time. It's instant. Maybe the algorithm itself can be optimized without resorting to open mp? Did you measure a noticeable improvement?

> Not sure what you mean, reSIDfp is a fork with "experimental" features like compile-time calculated combined waveforms and two-step resampling. All changes that have been proved good was merged back into reSID, currently maintained as part of VICE. So both emulations are quality-wise excellent with just some minor differences.

Yes, maybe I was too concerned about those differences making an audible/quality difference to users. I'm not too deep into the emulation.

There are tons more warnings that I have to suppress, because llvm/clang gets a lot better with diagnostics all the time, so with every new version, it shows you more and more where one unintentionally created UB (which is always dangerous).

**Reply by drfiemost** [reply]
> 1. So for the filter and resample tables, a simple mutex would be fine, but for the re-sampling table I would actually switch to a locally owned version, as this allows the sample-rate to be different between threads. This can be useful for exporting a e.g. 48 kHz WAV-file, while listening to a different tune at 44.1 kHz. 

The resample tables are cached for each samplerate (see the FIR_CACHE above), so you'll need a locking mechanism when building a new table for a samplerate not in cache.

> We also have to remember: as long as you only have one thread (and more is not even possible right now), the memory consumption would be the exact same. So what is there to lose? Making all these variables non-static doesn't lose anything. You're not spending any more memory, start-up time, or anything else. It stays the same. I truly don't understand why they are static in the first place. It provides zero benefits, at the cost of multi-threaded compatibility.

I don't see a reason for wasting MBs of memory for the same values and wasting CPU cycles to rebuild tables, but maybe I'm too old-fashioned ;)

> `-Wsource-uses-openmp` gets thrown by clang. It's because you've used `#pragma omp` outside the guards.

Not sure it's a meaningful warning, pragmas should be ignored by the compiler if unused.

> I can measure performance, but even in debug mode (which on my machine is more than 10x as slow as release, I don't notice any initialization time. It's instant. Maybe the algorithm itself can be optimized without resorting to open mp? Did you measure a noticeable improvement?

I don't recall the numbers, for the record I'm on an AMD FX-6300 from 2012, not exactly bleeding-edge...

> There are tons more warnings that I have to suppress, because llvm/clang gets a lot better with diagnostics all the time, so with every new version, it shows you more and more where one unintentionally created UB (which is always dangerous).

Will give llvm a spin, feel free to report suspicious warnings.

**Comment by Toddler-Boy** [comment]
> I don't see a reason for wasting MBs of memory for the same values and wasting CPU cycles to rebuild tables, but maybe I'm too old-fashioned ;)

Can you provide examples for your reasoning? Right now, it's impossible to even use libsidplayfp in a multithreaded manner. So what is gained by this caching? It's not like multiple instance could share these tables.

I find it weird that you argue memory savings, but then you have a map of caches, meaning if I switch the sample-rate from 44100 to 48000, I now have two tables in memory, even though I only need the one (the current sample-rate).

> Not sure it's a meaningful warning, pragmas should be ignored by the compiler if unused.

Of course it's meaningful. You are telling the compiler "here is a section for OpenMP" without having OpenMP enabled. It's telling you "hey, you obviously intend to use OpenMP here, but it's actually turned off". That's a warning. It's giving you a heads up that you might accidentally forgot to toggle a switch.

It's a similar principle how you nowadays have to provide the "override" keyword, so the compiler knows for sure you mean to override that function. It warns you if you forget, so you know the possible consequences.

> feel free to report suspicious warnings.

There are simply too many. I had to suppress the following warnings:

```
    #pragma clang diagnostic ignored "-Wshorten-64-to-32"
    #pragma clang diagnostic ignored "-Wnon-virtual-dtor"
    #pragma clang diagnostic ignored "-Wunused-private-field"
    #pragma clang diagnostic ignored "-Wimplicit-int-conversion"
    #pragma clang diagnostic ignored "-Wimplicit-float-conversion"
    #pragma clang diagnostic ignored "-Winconsistent-missing-destructor-override"
    #pragma clang diagnostic ignored "-Wsource-uses-openmp"
    #pragma clang diagnostic ignored "-Wshadow"
    #pragma clang diagnostic ignored "-Wshadow-field"
    #pragma clang diagnostic ignored "-Wfloat-conversion"
```

Each of these happens multiple, sometimes dozens, of times, in the code. I would gladly fix them, but I'm not sure that's wanted. It involves renaming local variables to avoid shadowing class member names (I usually add a single underscore to the local name, so ```sidNo``` becomes ```_sidNo```) so the compiler knows which variable (member or local) you actually want read/write/modify.

Lot's of implicit conversions. I would even replace ```short``` with ```int16_t``` for extra clarity.

**Reply by drfiemost** [reply]
> I find it weird that you argue memory savings, but then you have a map of caches, meaning if I switch the sample-rate from 44100 to 48000, I now have two tables in memory, even though I only need the one (the current sample-rate).

The cache was meant to help with multiple sequential plays, back in a single-threaded times. It needs to be adapted for today multi-threading environments.

> Of course it's meaningful. You are telling the compiler "here is a section for OpenMP" without having OpenMP enabled. It's telling you "hey, you obviously intend to use OpenMP here, but it's actually turned off". That's a warning. It's giving you a heads up that you might accidentally forgot to toggle a switch.

Ok, will check, but not high in my priority list.

> There are simply too many. I had to suppress the following warnings:
> 
> ```
>     #pragma clang diagnostic ignored "-Wshorten-64-to-32"
>     #pragma clang diagnostic ignored "-Wnon-virtual-dtor"
>     #pragma clang diagnostic ignored "-Wunused-private-field"
>     #pragma clang diagnostic ignored "-Wimplicit-int-conversion"
>     #pragma clang diagnostic ignored "-Wimplicit-float-conversion"
>     #pragma clang diagnostic ignored "-Winconsistent-missing-destructor-override"
>     #pragma clang diagnostic ignored "-Wsource-uses-openmp"
>     #pragma clang diagnostic ignored "-Wshadow"
>     #pragma clang diagnostic ignored "-Wshadow-field"
>     #pragma clang diagnostic ignored "-Wfloat-conversion"
> ```
> 
> Each of these happens multiple, sometimes dozens, of times, in the code. I would gladly fix them, but I'm not sure that's wanted. It involves renaming local variables to avoid shadowing class member names (I usually add a single underscore to the local name, so `sidNo` becomes `_sidNo`) so the compiler knows which variable (member or local) you actually want read/write/modify.

Well, the shorten-64-to-32 and destructor warnings are worth checking, maybe also the implicit conversion ones. I'd start with these.


> Lot's of implicit conversions. I would even replace `short` with `int16_t` for extra clarity.

I agree but it's used in lots of places. I hope MS compilers provide the stdint header nowadays.

**Reply by emoon** [reply]
>  I hope MS compilers provide the stdint header nowadays.

It does so it's fine to use.

**Comment by Toddler-Boy** [comment]
> The cache was meant to help with multiple sequential plays, back in a single-threaded times. It needs to be adapted for today multi-threading environments.

Fair enough, but why must it be declared "static"? What's the benefit of that? It doesn't do anything in single-threaded contexts, and in multi-threaded contexts, it's incompatible. Removing ALL "static" keywords from variables would be a big step in making this library multi-threaded save.

It won't affect single-threaded apps at all. Nothing would change for them.

> I agree but it's used in lots of places. I hope MS compilers provide the stdint header nowadays.

I understand where you're coming from, given MS track record, but it's important to understand that for the past few years, there has been a push towards modernization in the C++ space for them. Their compiler (when limited to C++20) is more feature-complete than the usual competition (clang, GCC, Intel, etc.)

Just look at this: https://en.cppreference.com/w/cpp/compiler_support

And scroll down to the C++20 section. Core language features and library features are ahead of everyone else. GCC is very close but lacks full "modules" support, which MSVC has.

For C++23 and C++26 it's different, but I don't think that's the issue here 😄

**Reply by drfiemost** [reply]
> Fair enough, but why must it be declared "static"? What's the benefit of that? It doesn't do anything in single-threaded contexts, and in multi-threaded contexts, it's incompatible. Removing ALL "static" keywords from variables would be a big step in making this library multi-threaded save.

It's declared static so it is shared among instances, it just lacks proper access control.

**Reply by Toddler-Boy** [reply]
Maybe it's my brain preventing me from understanding a use case for this logic, but I don't understand. What would these different instances be used for?

What would be the use case for multiple instances in a single-threaded design? Why not re-use the existing instance? Load a different tune into it, change the sample rate, etc. What good can a second instance do that the existing instance can not?

If you are truly 100% against removing the static keyword everywhere, then please let me know, so I can move on.

In that case, I'd probably fork this, make my changes, and then continuously merge your future changes into my fork.

**Reply by drfiemost** [reply]
Ok, but in a multi-threaded scenario you need a single static cache available to all instances.
Every new instance would lock the cache and ask for a resampling table given the sampling-rate, if it's ready it gets the pointer and releases the lock, otherwise it creates a new table and put it into the cache.
Similar thing for the filter tables, only the first instance is required to build the tables, the others can reuse them once ready.

Or are you thinking of a different solution?
Anyway feel free to fork, we can then discuss about merging back the changes if you will.

**Reply by Toddler-Boy** [reply]
> Ok, but in a multi-threaded scenario you need a single static cache available to all instances.

No, you don't. You simply live with the fact that each instance uses a couple of megabytes of memory for its tables. Otherwise, the locks would prevent other instances from finishing their instancing until the current instance is finished with its table calculations. You're effectively blocking everybody to save 2-3 MB of memory.

> Similar thing for the filter tables, only the first instance is required to build the tables, the others can reuse them once ready.

Even that would be created for each instance, so as not to block other instances.

> Or are you thinking of a different solution?

Yes, not using the static keyword anywhere and having each instance calculate its own tables.

> Anyway feel free to fork, we can then discuss about merging back the changes if you will.

I'm afraid you will find plenty of reasons not to merge back, as I'm also planning on modernizing a few other bits of code. All these iffy macros for unused parameters, etc., should really be replaced with ```[[ maybe_unsed ]]```, and all the shadowed variables need to be renamed. All class members should be initialized during declaration, etc.

My goal is to make this whole auto conf business obsolete. It's not needed anymore. You can test everything during compile-time.

I also find it completely unnecessary to check for external crypt libraries to use md5, etc. if a local, known as good, implementation is available. You're saving a few kilobytes of code here by having the option to turn the local version off, at the cost of build complexity. A bad tradeoff in my opinion.

**Reply by Toddler-Boy** [reply]
Oh, and I would also remove all the OpenMP stuff. The calculations are fast enough. On sufficiently fast machines, spawning multiple threads takes longer than just doing the calculations. This is due to OS overhead for thread management. It's additional complexity for no, or very questionable, gain.

**Reply by drfiemost** [reply]
I see your points. It is always good to have alternatives with different goals, so wish you happy hacking!

**Reply by Toddler-Boy** [reply]
I've created a fork here:

https://github.com/reFX/libsidplayfp

I've also created a branch called 'modernized'. I will try my best to merge any relevant changes from master into the modernized branch.

It is my sincere hope that you will come to accept modernity and eventually merge my branch into yours.

**Reply by Toddler-Boy** [reply]
Before I start rewriting code, I did a quick analysis:

Memory usage for Resampling tables:

1) From NTSC to intermediate frequency = 2,929 shorts = 5,858 BYTES = 5,8 KiloBytes. Kilo. Not mega.
2) From intermediate to output frequency (44.1 kHz) = 18,972 shorts = 37,944 BYTES = 37,9 KiloButes. Kilo. Not mega.

For PAL

1) 2,970 shorts = 5,940 byes = 5.9 KiloBytes.
2) 18,875 shorts = 37,750 bytes = 37,7 KiloBtes.

Replacing that cache with two simple ```std::vector<int16_t>``` is a no brainer to me. It costs literally no time to calculate, takes almost no memory and all that code creating hashes, etc. becomes obsolete.

For the ```FilterModelConfig6581``` the memory usage is a bit higher (10 MB, bytes, not shorts, doubles, etc. 10 MB is all that's needed), but even that is not a lot by today's standards.

This is my last ditch attempt to convince you to let us simply drop the usage of static variables/members. The savings are minimal (computationally and memory usage) and are not worth the pain. Code becomes simpler, easer to maintain, easier to understand and it becomes immediately multi-threaded friendly, without the need for mutexes.

**Reply by drfiemost** [reply]
I agree about the resampling tables being quite small, not so sure about the filter ones.

Anyway, the project is mature and stable and I'm not going to do any major change for the time being.
I might have a look at your work and cherry pick a few things here and there, depending on mood and free time.

**Reply by Toddler-Boy** [reply]
I've figured as much. In the meantime, I've decided to make my branch much more specialized and remove certain features. I can avoid a few virtual functions, remove a lot of wrappers and interface classes, and replace vectors with a size of three with simple arrays. It seems a lot of indirection was added during the conversion from C to Java and then from Java to C++.

BTW, one thing you could cherry-pick immediately would be my removal of OpenMP with std::thread. The startup time for both filter configs has been lowered from 300ms to 100ms (on my machine, in release mode). No defines needed, no compiler switches, no config file. It's a C++11 standard feature.

My specialized version uses residfp directly. No wrapper. No more exSID, HardSID, or reSID.
No more support for muting voices or fast-forwarding.

Muting voices has unintended side effects, as the filter doesn't get as much input anymore and thus doesn't distort anymore. Also, most tunes sound weird when voices are muted. If someone wants to analyze a tune for a cover version, they can use other tools.

Fast forwarding is so flawed, as once you're past the point accidentally, there is no way to rewind. Then, you have to start the whole song over again. It's also hell on the CPU. It's better to render the entire tune into RAM and provide the means to rewind/fast-forward using that data.

I'm aiming to provide a result that reflects the original author's intent as closely as possible.

**Comment by mywave82** [comment]
Making the library re-entrant will also remove weird corner-cases in general.

Multimedia programs might hold multiple object instance in memory in order to do cross-fading.

If a program changes output device, the playback rate (and other parameters) can change.
 
If a plugin that used libsidplayfp is unloaded (and all libsidplayfp related objects are deleted), should the cache be cleared to avoid unreferenced memory?

**Comment by Toddler-Boy** [comment]
> If a plugin that used libsidplayfp is unloaded (and all libsidplayfp related objects are deleted), should the cache be cleared to avoid unreferenced memory?

Isn't that automatic?
